### PR TITLE
#239 now `ipgeolocation` requests are cached

### LINKED ISSUE #239: ipgeolocation requests are not cached
Currently, we make requests to their server on every IP address. I believe, most IP addresses will be repeated (since they are all from Github.com). Let's cache them, using `settings.zache`, similar to how we cache other data.

Moreover, I would change the implementation we have now. Instead of making the HTTP request to ipgeolocation server while generating our HTML, we should render the ``. Then, we should define a new Sinatra entry point `/flag-of/.*` where we make a request to ipgeolocation server, download the flag from their server, and return its binary content.

### PR DESCRIPTION
**Pull request by Suban05**
Fixes #239

### CODE REVIEW

File: test/front/test_front_flag_of.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+require_relative '../../baza'
+
+class Baza::FrontFlagOfTest < Minitest::Test
+  def app
+    Sinatra::Application
+  end
+
+  def test_returns_image_as_binary_data
+    WebMock.disable_net_connect!
+    body = 'image data'
```
Comment (Yegorov) [inline test/front/test_front_flag_of.rb:36]: @Suban05 Maybe body should looks as binary?
```ruby
body = SecureRandom.bytes(100)
=> "\xDC/\x8DB\xF0]\x81\x89\xDAB\xD50\x1D\xFAw\xDB\x02\xFC{\x8C\xE4T\xBF\xD2\xD3\x9C\xED\x99\x19\x1F\xE3$T=j\xEF\xA2J1\x90\xA6*C\x17\x90\xE7\xAA\xE1\xAC\x04\x14j\x11\xA2\xA1\xD7\x96_v\xB8fFe+\xE3-\x80mz\xE4\xDBm\xF2#\x95+\x98G\xF7JP\xB6\x92\x8C\v\xD1p\x18\x18=\x9F\x00\xC0\xD0\xFBp\x89\xE9\xED^"
```

File: front/front_flag_of.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  if src.nil?
+    src = settings.ipgeolocation.ipgeo(ip:)['country_flag']
+    settings.zache.get(:ipgeolocation)[ip] = src
+  end
+  src
+end
+
+get(%r{/flag-of/(.*)}) do
+  content_type 'image/png'
+  URI.parse(path_to_flag(params['captures'].first, settings)).open.read
```
Comment (Yegorov) [inline front/front_flag_of.rb:38]: @Suban05 Why do you need to read the file and give it to the client, and not just give the link?

**Review by Yegorov — COMMENTED** [review]
@Suban05 Looks good, but there are several questions

**Comment by Suban05** [comment]
@Yegorov fixed, please check

File: front/front_flag_of.rb (hunk 2 above)
Comment (Yegorov) [reply front/front_flag_of.rb:35]: @Suban05 Okey, but it seems here we need some kind of caching

**Review by Yegorov — APPROVED** [review]
@Suban05 great!

**Comment by Suban05** [comment]
@yegor256 please check

File: front/front_flag_of.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+require 'open-uri'
+
+def path_to_flag(ip, settings)
```
Comment (yegor256) [inline front/front_flag_of.rb:27]: @Suban05 why do you need this function (in global scope)? Can't you inline it?

File: baza.rb [hunk 4]
```diff
@@ -305,5 +306,6 @@
 require_relative 'front/front_alterations'
 require_relative 'front/front_swarms'
 require_relative 'front/front_push'
+require_relative 'front/front_flag_of'
```
Comment (yegor256) [inline baza.rb:309]: @Suban05 maybe we can put this to `front_assets` or `front_misc`?

File: baza.rb [hunk 5]
```diff
@@ -229,6 +229,7 @@
     token:,
     connection: ENV['RACK_ENV'] == 'test' && token.empty? ? Baza::IpGeolocation::FakeConnection : Faraday
   )
+  set :ipgeolocation_cache, Zache.new
```
Comment (yegor256) [inline baza.rb:232]: @Suban05 why not use existing `settings.zache`?

File: front/front_flag_of.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  settings.ipgeolocation_cache.get(ip) do
+    settings.ipgeolocation.ipgeo(ip:)['country_flag']
+  end
+end
+
+get(%r{/flag-of/(.*)}) do
```
Comment (yegor256) [inline front/front_flag_of.rb:33]: @Suban05 maybe we can make the regex more exclusive, like `^/flag-of/[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$/`

**Comment by yegor256** [comment]
@Suban05 see a few comments above

File: baza.rb (hunk 5 above)
Comment (Suban05) [reply baza.rb:232]: @yegor256 I wanted to do it this way: `settings.zache.put(:ipgeolocation, {}) unless settings.zache.exists?(:ipgeolocation)`, but see this comment https://github.com/zerocracy/baza/pull/242#discussion_r1747502655
can we write it like this? 
```rb
settings.zache.get(:ipgeolocation) { Zache.new }
```
and next, we can do it like this: `settings.zache.get(:ipgeolocation).get(ip)`
what do you think?

Comment (yegor256) [reply baza.rb:232]: @Suban05 maybe just `settings.zache.get('country-of-#{ip}')` returning the name of the country (or the URL of the .png)

**Comment by Suban05** [comment]
@yegor256 please check

File: front/front_assets.rb [hunk 7]
```diff
@@ -60,3 +61,13 @@
     html:
   )
 end
+
+get(%r{/flag-of/([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})}) do
+  ip = params['captures'].first
+  content_type 'image/png'
+  src =
+    settings.zache.get("country-of-#{ip}") do
```
Comment (yegor256) [inline front/front_assets.rb:69]: @Suban05 maybe it's more accurate to call it `flag-of-#{ip}`?

**Comment by Suban05** [comment]
@yegor256 updated

**Comment by yegor256** [comment]
@Suban05 thanks!

**Comment by 0crat** [comment]
@Yegorov Hey there! 👋 Great job on the code review! You've snagged +4 points for your effort. That's pretty cool, considering we factor in things like hits-of-code and comment count. Your running balance is now [+330](https://zerocracy.github.io/judges-action/zerocracy-vitals.html) - keep up the awesome work! Remember, more detailed reviews with thoughtful comments can boost your points even higher. Looking forward to your next contribution! 💻🚀

**Comment by 0crat** [comment]
@Suban05 Hey there! 👋 Great job on your contribution! You've snagged +4 points this time around. Here's the breakdown: +4 base points, -4 due to the 21 comments during review (a bit chatty, huh?), and +4 to keep you going. Remember, [quality over quantity](https://www.yegor256.com/2018/03/06/speed-vs-quality.html) is key! Keep those contributions coming, but maybe aim for fewer comments next time. Your balance is now sitting pretty at [+337](https://zerocracy.github.io/judges-action/zerocracy-vitals.html). Keep up the good work!

**Pull request merged**